### ISSUE TITLE
slow speed in 1*1 small dense convolution

### ISSUE THREAD

**Issue by peteryang1** [intel/mkl-dnn#437]
I found the 1x1 dense conv in MKLDNN is very slow in small node!

to prove that, I write a mkl version of 1x1 dense conv(just use one gemm to calculate the output, if stride != 1, I will put the input's useful data into a workspace and do the gemm) and make a report about the speed of 1x1 conv node. Here is what I get:

| input size   | kernel size | stride | pad       | group | alogrithm     |   Time   |
|--------------|-------------|--------|-----------|-------|---------------|--------|
| [1,16,60,60] | [16,16,1,1] | [1,1]  | [0,0,0,0] | 1     | MKLML GEMM    | 70.72  |
| [1,16,60,60] | [16,16,1,1] | [1,1]  | [0,0,0,0] | 1     | MKLDNN Direct | 140.43 |
| [1,16,60,60] | [16,16,1,1] | [2,2]  | [0,0,0,0] | 1     | MKLML GEMM    | 32.96  |
| [1,16,60,60] | [16,16,1,1] | [2,2]  | [0,0,0,0] | 1     | MKLDNN Direct | 182.52 |

time above only contains runtime, in MKLDNN it only contains the code
```C++
    auto conv = mkldnn::convolution_forward(
        pdd->conv_primitive_desc, input_memory, kernel_memory, bias_memory,
        output_memory);

    mkldnn::impl::event_t e;
    conv.get()->execute(&e);
```

It seems even a very small conv node could cause above 100 us. 

I'm a little confused why MKLDNN costs so much time in this small conv node, and is there a chance I can improve that?
I'm running the release 0.18.1 and cpu is Intel i7-4790.

Thanks!

**Comment by nathan-greeneltch-intel** [comment]
There are overhead costs for the first convolution in a topology. You can test this by running a single conv as a proxy for an init method, then underneath write a for-loop that runs the same convolution 1k times. Only time the for-loop.

**Comment by tprimak** [comment]
Hi @peteryang1! Could you please check MKL-DNN convolution implementation that was executed in your case? You can do it by running your example with MKLDNN_VERBOSE=1

**Comment by peteryang1** [comment]
thanks @nathan-greeneltch-intel  @tprimak for your help!
for @nathan-greeneltch-intel : I thought about this ahead, so I did something to prevent it:
1. this chart is a very small part of a very big chart, I run the input from (1,16,10,10) to (1,16,100,100), 60x60 is in the middle of the chart(by the way, smaller input like 1x16x10x10 mkldnn is way more slower);
2. this time is an average of 10 consecutive run;

for @tprimak : 
I tested it, it uses `jit_avx2_1x1_convolution_fwd_t` implementation.

**Comment by tprimak** [comment]
Hi @peteryang1  Do you have hyper-threading ON or OF on your machine? Do you set OMP_NUM_THREADS and KMP_AFFINITY for your run? 

When do benchmarking, do you remove the timing of the first "warm up" iteration? Can you check if your time for direct convolution is the same as benchdnn reports for corresponding shapes?

**Comment by peteryang1** [comment]
Hi, @tprimak 
I had threading ON on my computer, and because I need to stay in single thread to avoid thread switching in multi-thread, so I disabled openmp when building.
I do throw the first time in my test, speed keeps the same. And I tried benchdnn, I'm a little confused, the stride 1*1 showed really fast, but stride 2*2 is really slow, and I found that in benchdnn. stride 2*2 uses gemm instead avx2. here is my report:
``` C++
.\benchdnn.exe --conv --mode=p g1mb1ic16ih65iw65oc16oh65
ow65kh1kw1sh1sw1ph0pw0n"conv"
perf,conv,g1mb1ic16ih65oc16oh65kh1nconv,0.0021632,0,0.029365,73.6659,0.0298477,72.4745
tests:1 passed:0 skipped:0 mistrusted:0 unimplemented:0 failed:0
total perf: min(ms):0.029365 avg(ms):0.0298477
 .\benchdnn.exe --conv --mode=p g1mb1ic16ih65iw65oc16oh65
ow65kh1kw1sh2sw2ph0pw0n"conv"
perf,conv,g1mb1ic16ih65oc16oh65kh1sh2ph32nconv,0.000557568,0,0.16593,3.36026,0.168415,3.31068
tests:1 passed:0 skipped:0 mistrusted:0 unimplemented:0 failed:0
total perf: min(ms):0.16593 avg(ms):0.168415
```

**Comment by peteryang1** [comment]
I got the mkldnn verbose report in my test, it does almost fit the result in my report.
```C++
Stride:1x1
mkldnn_verbose,create,reorder,jit:uni,undef,in:f32_goihw out:f32_gOIhw8i8o,num:1,1x16x16x1x1,0.0401995
mkldnn_verbose,create,convolution,jit_1x1:avx2,forward_inference,fsrc:nChw8c fwei:gOIhw8i8o fbia:x fdst:nChw8c,alg:convolution_direct,mb1_g1ic16oc16_ih60oh60kh1sh1dh0ph0_iw60ow60kw1sw1dw0pw0,0.0869565
mkldnn_verbose,create,reorder,jit:uni,undef,in:f32_goihw out:f32_gOIhw8i8o,num:1,1x16x16x1x1,0.0467569
mkldnn_verbose,create,convolution,jit_1x1:avx2,forward_inference,fsrc:nChw8c fwei:gOIhw8i8o fbia:x fdst:nChw8c,alg:convolution_direct,mb1_g1ic16oc16_ih60oh60kh1sh1dh0ph0_iw60ow60kw1sw1dw0pw0,0.0712758

Stride 2x2:
mkldnn_verbose,create,reorder,jit:uni,undef,in:f32_goihw out:f32_gOIhw8i8o,num:1,1x16x16x1x1,0.0433357
mkldnn_verbose,create,convolution,jit_1x1:avx2,forward_inference,fsrc:nChw8c fwei:gOIhw8i8o fbia:x fdst:nChw8c,alg:convolution_direct,mb1_g1ic16oc16_ih60oh30kh1sh2dh0ph0_iw60ow30kw1sw2dw0pw0,0.103207
mkldnn_verbose,create,reorder,jit:uni,undef,in:f32_goihw out:f32_gOIhw8i8o,num:1,1x16x16x1x1,0.0427655
mkldnn_verbose,create,convolution,jit_1x1:avx2,forward_inference,fsrc:nChw8c fwei:gOIhw8i8o fbia:x fdst:nChw8c,alg:convolution_direct,mb1_g1ic16oc16_ih60oh30kh1sh2dh0ph0_iw60ow30kw1sw2dw0pw0,0.102352
```

**Comment by peteryang1** [comment]
I think I found the cause of the problem, in each run, I will create the primitive and then run, so my execution time includes create and submit.

I checked the code before, thinking that creating the primitive does not cost a lot of time, but seems in this kind of small node, this period of time matters a lot.

I should have found the problem when I firstly pasted the code in the beginning, and I will try to deal with it, thanks @nathan-greeneltch-intel @tprimak for all your help!
